File: Code/Projects/PageValidator/PyPageValidator/Validator/validator.py
from __future__ import annotations
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import List, Set, Tuple

sys.path.insert(0, str(Path(__file__).parent.parent))
from Lexer.lexer import (
    Lexer, Attr, OpenTag, SelfClosingTag, CloseTag,
    TextNode, CommentNode, DoctypeDecl,
)


# ---------------------------------------------------------------------------
# ValidationError — one rule violation found in an HTML file.
# ---------------------------------------------------------------------------

@dataclass
class ValidationError:
    rule: str
    message: str
    line: int
    col: int


# ---------------------------------------------------------------------------
# Report — the complete result for a single file.
# ---------------------------------------------------------------------------

class Report:
    def __init__(self, file: str, errors: List[ValidationError]) -> None:
        self.file = file
        self.errors: List[ValidationError] = errors

    @property
    def is_valid(self) -> bool:
        return len(self.errors) == 0


# ---------------------------------------------------------------------------
# Validator — stateless; all state lives in validate()'s locals.
# ---------------------------------------------------------------------------

_VOID_ELEMENTS: frozenset[str] = frozenset({
    'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input',
    'link', 'meta', 'param', 'source', 'track', 'wbr',
})


class Validator:
    @staticmethod
    def validate(src: str, file: str) -> Report:
        lexer = Lexer(src)
        errors: List[ValidationError] = []
        stack: List[Tuple[str, Tuple[int, int]]] = []
        ids: Set[str] = set()

        seen_doctype = False
        html_count = 0
        seen_head = False
        seen_title = False
        seen_body = False
        in_head = False

        lexeme = lexer.next_lexeme()
        while lexeme is not None:
            if isinstance(lexeme, DoctypeDecl):
                seen_doctype = True

            elif isinstance(lexeme, OpenTag):
                _check_attrs(lexeme.attrs, lexeme.pos, errors, ids)
                if lexeme.name == 'html':
                    html_count += 1
                elif lexeme.name == 'head':
                    seen_head = True
                    in_head = True
                elif lexeme.name == 'title' and in_head:
                    seen_title = True
                elif lexeme.name == 'body':
                    seen_body = True
                if lexeme.name not in _VOID_ELEMENTS:
                    stack.append((lexeme.name, lexeme.pos))

            elif isinstance(lexeme, SelfClosingTag):
                _check_attrs(lexeme.attrs, lexeme.pos, errors, ids)

            elif isinstance(lexeme, CloseTag):
                if lexeme.name in _VOID_ELEMENTS:
                    errors.append(ValidationError(
                        'void-elements',
                        f'void element <{lexeme.name}> must not have a close tag',
                        lexeme.pos[0], lexeme.pos[1],
                    ))
                else:
                    if lexeme.name == 'head':
                        in_head = False
                    if stack and stack[-1][0] == lexeme.name:
                        stack.pop()
                    elif stack:
                        errors.append(ValidationError(
                            'tag-nesting',
                            f'</{lexeme.name}> does not match open <{stack[-1][0]}>',
                            lexeme.pos[0], lexeme.pos[1],
                        ))
                    else:
                        errors.append(ValidationError(
                            'tag-nesting',
                            f'</{lexeme.name}> has no matching open tag',
                            lexeme.pos[0], lexeme.pos[1],
                        ))

            lexeme = lexer.next_lexeme()

        # Post-loop structural checks
        if not seen_doctype:
            errors.append(ValidationError('doctype', 'document is missing <!DOCTYPE html>', 1, 1))
        if html_count != 1:
            errors.append(ValidationError(
                'root-element',
                f'expected exactly 1 <html> element, found {html_count}', 1, 1,
            ))
        if not seen_head:
            errors.append(ValidationError('head-required', 'document is missing a <head> element', 1, 1))
        elif not seen_title:
            errors.append(ValidationError('head-required', '<head> is missing a <title> element', 1, 1))
        if not seen_body:
            errors.append(ValidationError('body-required', 'document is missing a <body> element', 1, 1))
        for name, pos in stack:
            errors.append(ValidationError(
                'tag-nesting',
                f'<{name}> was opened but never closed',
                pos[0], pos[1],
            ))

        return Report(file, errors)


def _check_attrs(
    attrs: List[Attr],
    pos: Tuple[int, int],
    errors: List[ValidationError],
    ids: Set[str],
) -> None:
    for attr in attrs:
        if not attr.quoted and attr.value:
            errors.append(ValidationError(
                'attr-quotes',
                f"attribute '{attr.key}' value '{attr.value}' is not quoted",
                pos[0], pos[1],
            ))
        if attr.key == 'id' and attr.value:
            if attr.value in ids:
                errors.append(ValidationError(
                    'duplicate-id',
                    f"duplicate id '{attr.value}'",
                    pos[0], pos[1],
                ))
            else:
                ids.add(attr.value)

File: Tools/CodeMetrics/PageValidator/CsPageValidator/Validator/Validator.cs
namespace PageValidator;

// ---------------------------------------------------------------------------
// ValidationError — one rule violation found in an HTML file.
// ---------------------------------------------------------------------------

public sealed record ValidationError(string Rule, string Message, int Line, int Col);

// ---------------------------------------------------------------------------
// Report — the complete result for a single file.
// ---------------------------------------------------------------------------

public sealed class Report
{
    public string File { get; }
    public IReadOnlyList<ValidationError> Errors { get; }
    public bool IsValid => Errors.Count == 0;

    public Report(string file, IReadOnlyList<ValidationError> errors)
    {
        File   = file;
        Errors = errors;
    }
}

// ---------------------------------------------------------------------------
// Validator — stateless; all state lives in Validate()'s locals.
// ---------------------------------------------------------------------------

public static class Validator
{
    private static readonly HashSet<string> VoidElements =
        new(StringComparer.OrdinalIgnoreCase)
        {
            "area", "base", "br", "col", "embed", "hr", "img", "input",
            "link", "meta", "param", "source", "track", "wbr",
        };

    public static Report Validate(string src, string file)
    {
        var lexer  = new Lexer(src);
        var errors = new List<ValidationError>();
        // List used as stack so remaining items iterate in opening order.
        var stack  = new List<(string Name, (int Line, int Col) Pos)>();
        var ids    = new HashSet<string>(StringComparer.Ordinal);

        bool seenDoctype = false;
        int  htmlCount   = 0;
        bool seenHead    = false;
        bool seenTitle   = false;
        bool seenBody    = false;
        bool inHead      = false;

        Lexeme? lexeme;
        while ((lexeme = lexer.NextLexeme()) is not null)
        {
            switch (lexeme)
            {
                case DoctypeDecl:
                    seenDoctype = true;
                    break;

                case OpenTag ot:
                    CheckAttrs(ot.Attrs, ot.Pos, errors, ids);
                    switch (ot.Name)
                    {
                        case "html":  htmlCount++; break;
                        case "head":  seenHead = true; inHead = true; break;
                        case "title" when inHead: seenTitle = true; break;
                        case "body":  seenBody = true; break;
                    }
                    if (!VoidElements.Contains(ot.Name))
                        stack.Add((ot.Name, ot.Pos));
                    break;

                case SelfClosingTag st:
                    CheckAttrs(st.Attrs, st.Pos, errors, ids);
                    break;

                case CloseTag ct:
                    if (VoidElements.Contains(ct.Name))
                    {
                        errors.Add(new ValidationError("void-elements",
                            $"void element <{ct.Name}> must not have a close tag",
                            ct.Pos.Line, ct.Pos.Col));
                        break;
                    }
                    if (ct.Name == "head") inHead = false;
                    if (stack.Count > 0 && stack[^1].Name == ct.Name)
                        stack.RemoveAt(stack.Count - 1);
                    else if (stack.Count > 0)
                        errors.Add(new ValidationError("tag-nesting",
                            $"</{ct.Name}> does not match open <{stack[^1].Name}>",
                            ct.Pos.Line, ct.Pos.Col));
                    else
                        errors.Add(new ValidationError("tag-nesting",
                            $"</{ct.Name}> has no matching open tag",
                            ct.Pos.Line, ct.Pos.Col));
                    break;
            }
        }

        // Post-loop structural checks
        if (!seenDoctype)
            errors.Add(new ValidationError("doctype",
                "document is missing <!DOCTYPE html>", 1, 1));

        if (htmlCount != 1)
            errors.Add(new ValidationError("root-element",
                $"expected exactly 1 <html> element, found {htmlCount}", 1, 1));

        if (!seenHead)
            errors.Add(new ValidationError("head-required",
                "document is missing a <head> element", 1, 1));
        else if (!seenTitle)
            errors.Add(new ValidationError("head-required",
                "<head> is missing a <title> element", 1, 1));

        if (!seenBody)
            errors.Add(new ValidationError("body-required",
                "document is missing a <body> element", 1, 1));

        foreach (var (name, pos) in stack)
            errors.Add(new ValidationError("tag-nesting",
                $"<{name}> was opened but never closed", pos.Line, pos.Col));

        return new Report(file, errors);
    }

    private static void CheckAttrs(
        IReadOnlyList<Attr> attrs,
        (int Line, int Col) pos,
        List<ValidationError> errors,
        HashSet<string> ids)
    {
        foreach (var attr in attrs)
        {
            if (!attr.Quoted && attr.Value.Length > 0)
                errors.Add(new ValidationError("attr-quotes",
                    $"attribute '{attr.Key}' value '{attr.Value}' is not quoted",
                    pos.Line, pos.Col));

            if (attr.Key == "id" && attr.Value.Length > 0)
            {
                if (!ids.Add(attr.Value))
                    errors.Add(new ValidationError("duplicate-id",
                        $"duplicate id '{attr.Value}'", pos.Line, pos.Col));
            }
        }
    }
}
